src/System.Console/src/System/IO/StdInStreamReader.cs

```diff
@@ -0,0 +1,259 @@
+                    x = '\0'; // Non-printable character.
+                    return ConsoleKey.Backspace;
+
+                case '\t':
+                    x = '\0'; // Non-printable character.
+                    return ConsoleKey.Tab;
+
+                case '\n':
+                    x = '\0'; // Non-printable character.
+                    return ConsoleKey.Enter;
+
+                case (char)(int)(0x1B):
+                    x = '\0'; // Non-printable character.
+                    return ConsoleKey.Escape;
+
+                case '*':
+                    return ConsoleKey.Multiply;
+
+                case '+':
+                    return ConsoleKey.Add;
+
+                case '-':
+                    return ConsoleKey.Subtract;
+
+                case '/':
+                    return ConsoleKey.Divide;
+
+                case (char)(int)(0x7F):
+                    x = '\0'; // Non-printable character.
+                    return ConsoleKey.Delete;
+
+                case ' ':
+                    return ConsoleKey.Spacebar;
+
+                default:
+                    // 1. Ctrl A to Ctrl Z.
+                    if (x >= 1 && x <= 26)
+                    {
+                        isCtrl = true;
+                        return ConsoleKey.A + x - 1;
+                    }
+
+                    // 2. Numbers from 0 to 9.
+                    if (x >= '0' && x <= '9')
+                    {
+                        return ConsoleKey.D0 + x - '0';
+                    }
+
+                    //3. A to Z
+                    if (x >= 'A' && x <= 'Z')
+                    {
+                        isShift = true;
+                        return ConsoleKey.A + (x - 'A');
+                    }
+
+                    // 4. a to z.
+                    if (x >= 'a' && x <= 'z')
+                    {
+                        return ConsoleKey.A + (x - 'a');
+                    }
+
+                    break;
+            }
+
+            return default(ConsoleKey);
+        }
+
+        internal bool MapBufferToConsoleKey(out ConsoleKey key, out char ch, out bool isShift, out bool isAlt, out bool isCtrl)
+        {
+            Debug.Assert(!IsExtraBufferEmpty());
+
+            isAlt = isCtrl = isShift = false;
+            int keyLength;
+
+            // 1. Try to get the special key match from the TermInfo static information.
+            if (ConsolePal.TryGetSpecialConsoleKey(_unprocessedBufferToBeRead, _startIndex, _endIndex, out key, out keyLength))
+            {
+                _startIndex += keyLength;
+                ch = '\0'; //Special keys are non-printable.
+                return true;
+            }
+
+            // 1. Check if we can match Esc + combination and guess if alt was pressed.
+            if (isAlt == false && _unprocessedBufferToBeRead[_startIndex] == (char)(int)(0x1B) /*Alt is send as an escape character*/ && _endIndex - _startIndex + 1 >= 2 /*We have at least two characters to read.*/)
```

Comment:
Sure, the only reason I prefer not to split these is because once split VS auto-formatting becomes a no-op and we get a ton of formatting issues in general :).